File: banking_tools/account.py
"""Account management module for banking operations."""

import json
import hashlib
from datetime import datetime
from typing import List, Dict, Optional


class Account:
    """Represents a bank account with basic operations."""
    
    def __init__(self, account_number: str, account_holder: str, 
                 password: str, initial_balance: float = 0.0):
        """
        Initialize a new bank account.
        
        Args:
            account_number: Unique account identifier
            account_holder: Name of the account holder
            password: Account password (will be hashed)
            initial_balance: Starting balance (default: 0.0)
        """
        self.account_number = account_number
        self.account_holder = account_holder
        self.password_hash = self._hash_password(password)
        self.balance = initial_balance
        self.transactions: List[Dict] = []
        self.created_at = datetime.now().isoformat()
        
        if initial_balance > 0:
            self._add_transaction("Initial Deposit", initial_balance, "credit")
    
    @staticmethod
    def _hash_password(password: str) -> str:
        """Hash password using SHA-256."""
        return hashlib.sha256(password.encode()).hexdigest()
    
    def verify_password(self, password: str) -> bool:
        """Verify if the provided password is correct."""
        return self.password_hash == self._hash_password(password)
    
    def _add_transaction(self, description: str, amount: float, 
                        transaction_type: str) -> None:
        """Add a transaction to the account history."""
        transaction = {
            "date": datetime.now().isoformat(),
            "description": description,
            "amount": amount,
            "type": transaction_type,
            "balance_after": self.balance
        }
        self.transactions.append(transaction)
    
    def deposit(self, amount: float, description: str = "Deposit") -> bool:
        """
        Deposit money into the account.
        
        Args:
            amount: Amount to deposit
            description: Transaction description
            
        Returns:
            True if successful, False otherwise
        """
        if amount <= 0:
            return False
        
        self.balance += amount
        self._add_transaction(description, amount, "credit")
        return True
    
    def withdraw(self, amount: float, description: str = "Withdrawal") -> bool:
        """
        Withdraw money from the account.
        
        Args:
            amount: Amount to withdraw
            description: Transaction description
            
        Returns:
            True if successful, False otherwise
        """
        if amount <= 0 or amount > self.balance:
            return False
        
        self.balance -= amount
        self._add_transaction(description, amount, "debit")
        return True
    
    def get_balance(self) -> float:
        """Get current account balance."""
        return self.balance
    
    def get_transaction_history(self, limit: Optional[int] = None) -> List[Dict]:
        """
        Get transaction history.
        
        Args:
            limit: Maximum number of transactions to return (None for all)
            
        Returns:
            List of transaction dictionaries
        """
        if limit:
            return self.transactions[-limit:]
        return self.transactions
    
    def to_dict(self) -> Dict:
        """Convert account to dictionary for serialization."""
        return {
            "account_number": self.account_number,
            "account_holder": self.account_holder,
            "password_hash": self.password_hash,
            "balance": self.balance,
            "transactions": self.transactions,
            "created_at": self.created_at
        }
    
    @classmethod
    def from_dict(cls, data: Dict) -> 'Account':
        """Create account from dictionary."""
        account = cls.__new__(cls)
        account.account_number = data["account_number"]
        account.account_holder = data["account_holder"]
        account.password_hash = data["password_hash"]
        account.balance = data["balance"]
        account.transactions = data["transactions"]
        account.created_at = data["created_at"]
        return account

File: src/main/java/com/banking/Account.java
package com.banking;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

/**
 * Represents a bank account with basic operations.
 */
public class Account {
    private String accountNumber;
    private String accountHolder;
    private String passwordHash;
    private double balance;
    private List<Transaction> transactions;
    private String createdAt;

    /**
     * Create a new bank account.
     * 
     * @param accountNumber Unique account identifier
     * @param accountHolder Name of the account holder
     * @param password Account password (will be hashed)
     * @param initialBalance Starting balance
     * @throws IllegalArgumentException if password is too short
     */
    public Account(String accountNumber, String accountHolder, String password, double initialBalance) {
        if (password.length() < 4) {
            throw new IllegalArgumentException("Password must be at least 4 characters long");
        }

        this.accountNumber = accountNumber;
        this.accountHolder = accountHolder;
        this.passwordHash = hashPassword(password);
        this.balance = initialBalance;
        this.transactions = new ArrayList<>();
        this.createdAt = LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);

        if (initialBalance > 0) {
            addTransaction("Initial Deposit", initialBalance, "credit");
        }
    }

    /**
     * Private constructor for deserialization.
     */
    private Account() {
        this.transactions = new ArrayList<>();
    }

    /**
     * Hash password using SHA-256.
     * 
     * @param password Plain text password
     * @return Hashed password
     */
    private static String hashPassword(String password) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(password.getBytes(StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 algorithm not found", e);
        }
    }

    /**
     * Verify if the provided password is correct.
     * 
     * @param password Password to verify
     * @return true if password matches, false otherwise
     */
    public boolean verifyPassword(String password) {
        return this.passwordHash.equals(hashPassword(password));
    }

    /**
     * Add a transaction to the account history.
     * 
     * @param description Transaction description
     * @param amount Transaction amount
     * @param type Transaction type (credit or debit)
     */
    private void addTransaction(String description, double amount, String type) {
        String date = LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
        Transaction transaction = new Transaction(date, description, amount, type, this.balance);
        this.transactions.add(transaction);
    }

    /**
     * Deposit money into the account.
     * 
     * @param amount Amount to deposit
     * @param description Transaction description
     * @return true if successful, false otherwise
     */
    public boolean deposit(double amount, String description) {
        if (amount <= 0) {
            return false;
        }

        this.balance += amount;
        addTransaction(description, amount, "credit");
        return true;
    }

    /**
     * Withdraw money from the account.
     * 
     * @param amount Amount to withdraw
     * @param description Transaction description
     * @return true if successful, false otherwise
     */
    public boolean withdraw(double amount, String description) {
        if (amount <= 0 || amount > this.balance) {
            return false;
        }

        this.balance -= amount;
        addTransaction(description, amount, "debit");
        return true;
    }

    /**
     * Get current account balance.
     * 
     * @return Current balance
     */
    public double getBalance() {
        return this.balance;
    }

    /**
     * Get transaction history.
     * 
     * @param limit Maximum number of transactions to return (null for all)
     * @return List of transactions
     */
    public List<Transaction> getTransactionHistory(Integer limit) {
        if (limit != null && limit > 0) {
            int size = transactions.size();
            int fromIndex = Math.max(0, size - limit);
            return new ArrayList<>(transactions.subList(fromIndex, size));
        }
        return new ArrayList<>(transactions);
    }

    // Getters
    public String getAccountNumber() {
        return accountNumber;
    }

    public String getAccountHolder() {
        return accountHolder;
    }

    public String getPasswordHash() {
        return passwordHash;
    }

    public List<Transaction> getTransactions() {
        return transactions;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    // Setters for deserialization
    public void setAccountNumber(String accountNumber) {
        this.accountNumber = accountNumber;
    }

    public void setAccountHolder(String accountHolder) {
        this.accountHolder = accountHolder;
    }

    public void setPasswordHash(String passwordHash) {
        this.passwordHash = passwordHash;
    }

    public void setBalance(double balance) {
        this.balance = balance;
    }

    public void setTransactions(List<Transaction> transactions) {
        this.transactions = transactions;
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }

    /**
     * Inner class representing a transaction.
     */
    public static class Transaction {
        private String date;
        private String description;
        private double amount;
        private String type;
        private double balanceAfter;

        public Transaction(String date, String description, double amount, String type, double balanceAfter) {
            this.date = date;
            this.description = description;
            this.amount = amount;
            this.type = type;
            this.balanceAfter = balanceAfter;
        }

        // Getters
        public String getDate() {
            return date;
        }

        public String getDescription() {
            return description;
        }

        public double getAmount() {
            return amount;
        }

        public String getType() {
            return type;
        }

        public double getBalanceAfter() {
            return balanceAfter;
        }

        // Setters for Gson
        public void setDate(String date) {
            this.date = date;
        }

        public void setDescription(String description) {
            this.description = description;
        }

        public void setAmount(double amount) {
            this.amount = amount;
        }

        public void setType(String type) {
            this.type = type;
        }

        public void setBalanceAfter(double balanceAfter) {
            this.balanceAfter = balanceAfter;
        }
    }
}
